python/mxnet/text/glossary.py

```diff
@@ -0,0 +1,654 @@
+
+
+        Parameters
+        ----------
+        embeds : mxnet.text.Embedding or list of mxnet.text.Embedding. If it
+            is a list of mxnet.text.Embedding instances, their embedding vectors
+            are concatenated for each token.
+        """
+
+        # Sanity check.
+        if isinstance(embeds, list):
+            for loaded_embed in embeds:
+                assert isinstance(loaded_embed, Embedding)
+        else:
+            assert isinstance(embeds, Embedding)
+            embeds = [embeds]
+
+        self._vec_len = sum(embed.vec_len for embed in embeds)
+        self._idx_to_vec = nd.zeros(shape=(len(self), self.vec_len))
+
+        col_start = 0
+        # Concatenate all the embedding vectors in embeds.
+        for embed in embeds:
+            col_end = col_start + embed.vec_len
+            self._idx_to_vec[:, col_start:col_end] = embed[self.idx_to_token]
+            col_start = col_end
+
+    # tokens is a str or a list of strs.
+    def update_idx_to_vec(self, tokens, new_vectors):
+        """Updates embedding vectors for tokens.
+
+
+        Parameters
+        ----------
+        tokens : str or a list of strs.
+            A token or a list of tokens whose embedding vector are to be
+            updated.
+        new_vectors : mxnet.ndarray.NDArray
+            A 2-D NDArray to be assigned to the embedding vectors of `tokens`.
+            Its length must be equal to the number of `tokens` and its width
+            must be equal to the dimension of embeddings of the glossary.
+        """
+
+        assert self.idx_to_vec is not None, \
+            'mxnet.text.Glossary._idx_to_vec has not been initialized.  Use ' \
+            'mxnet.text.Glossary.__init__() or ' \
+            'mxnet.text.Glossary.set_idx_to_embed() to initialize it.'
+
+        if not isinstance(tokens, list):
+            tokens = [tokens]
+
+        assert isinstance(new_vectors, nd.NDArray) and \
+               len(new_vectors.shape) == 2, 'new_vectors must be a 2-D NDArray.'
+        assert new_vectors.shape[0] == len(tokens), \
+            'The length of new_vectors must be equal to the number of tokens.'
+        assert new_vectors.shape[1] == self.vec_len, \
+            'The width of new_vectors must be equal to the dimension of ' \
+            'embeddings of the glossary.'
+
+        indices = []
+        for token in tokens:
+            if token in self.token_to_idx:
+                indices.append(self.token_to_idx[token])
+            else:
+                raise ValueError('Token %s is unknown to the glossary. To '
+                                 'update the embedding vector for an unknown '
+                                 'token, please specify it explicitly as the '
+                                 'unknown special token %s in tokens. This is '
+                                 'to avoid unintended updates.' %
+                                 (token, self.idx_to_token[Glossary.unk_idx()]))
```

Comment:
It’s a common to use one set (i.e. training set) for generating the vocabulary and reuse the same vocabulary on another set for indexing. Returning the index for unknown and warning the user would likely make this interface easier to use.

Reply:
Thanks. Since update_idx_to_vec is a member function of a specific Glossary instance, it is fair to assume that the input tokens match the token indices of this specific Glossary instance.

Reply:
OK